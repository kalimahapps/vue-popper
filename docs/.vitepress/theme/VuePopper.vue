<template>
	<VuePopperComponent hover>
		<template #default>
			<button>{{ text }}</button>
		</template>

		<template #content>
			<div>
				<div class="tooltip-text">
					<slot name="default">
						Hi, from vue-popper! üçø
					</slot>
				</div>
			</div>
		</template>
	</VuePopperComponent>
</template>

<script setup lang="ts">
import VuePopperComponent from '@kalimahapps/vue-popper';

defineProps({
	text: {
		type: String,
		default: 'Button',
	},
});
</script>

<style lang="scss">
.popper-container {
	.tooltip-text {
		font-size: 0.8rem;
		white-space: nowrap;
	}

	button {
		border-color: #21333c;
		background-color: #21333c;
		color: #ececec;
		padding: 0.6em 1.5em;
		transition: all 0.1s linear;
		&:hover {
			background-color: #2c4a57;
		}
	}

	.trigger-element {
		display: inline-block;
	}
}
</style>
