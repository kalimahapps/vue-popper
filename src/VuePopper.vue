<template>
	<div class="popper-container">
		<component :is="triggerElement">
			<slot />
		</component>

		<component :is="tooltipElement">
			<slot name="content" />
		</component>
	</div>
</template>

<script setup lang="ts">
import type { PropType } from 'vue';
import { computed } from 'vue';
import type { PositioningStrategy, Placement, Modifier } from '@popperjs/core';
import { useVuePopper } from './composable';

const properties = defineProps({
	/**
	 * Trigger element on hover
	 */
	hover: {
		type: Boolean,
		default: false,
	},

	/**
	 * Disable clicking outside to close
	 */
	disableClickOutside: {
		type: Boolean,
		default: false,
	},

	/**
	 * How many milliseconds to wait before opening the tooltip
	 */
	openDelay: {
		type: Number,
		default: 0,
		validator: (value: number) => {
			const isPositive = value >= 0;
			if (!isPositive) {
				console.warn('@kalimahapps/vue-popper: openDelay must be a positive number');
				return false;
			}
			return true;
		},
	},

	/**
	 * How many milliseconds to wait before closing the tooltip
	 */
	closeDelay: {
		type: Number,
		default: 0,
		validator: (value: number) => {
			const isPositive = value >= 0;
			if (!isPositive) {
				console.warn('@kalimahapps/vue-popper: closeDelay must be a positive number');
				return false;
			}

			return true;
		},
	},

	/**
	 * Display an arrow on the tooltip
	 */
	showArrow: {
		type: Boolean,
		default: true,
	},

	/**
	 * Whether to interact with tooltip when hover is true
	 */
	interactive: {
		type: Boolean,
		default: false,
	},

	/**
	 * Set the strategy used to position the popper
	 *
	 * @values absolute | fixed
	 */
	strategy: {
		type: String as PropType<PositioningStrategy>,
		default: 'absolute',
	},

	/**
	 * Set the placement of the popper
	 */
	placement: {
		type: String as PropType<Placement>,
		default: 'top',
	},

	/**
	 * Set the modifiers of the popper
	 */
	modifiers: {
		type: Array as PropType<Modifier<any, any>[]>,
		default: () => {
			return [];
		},
	},

	/**
	 * Set the type of animation
	 *
	 * @values none, fade, shift-in, shift-out, scale, CUSTOM
	 */
	animation: {
		type: String,
		default: 'shift-in',
	},

	/**
	 * Animation duration in milliseconds
	 */
	animationDuration: {
		type: [Number, Array],
		default: 300,
		validator(value) {
			if (Array.isArray(value)) {
				return value.every((durationValue) => {
					return typeof durationValue === 'number' && durationValue >= 0;
				});
			}
			return typeof value === 'number' && value >= 0;
		},
	},

	/**
	 * Popper options to merge with default options
	 *
	 * @see https://popper.js.org/docs/v2/
	 */
	popperOptions: {
		type: Object,
		default: () => { return {}; },
	},
});

const emit = defineEmits(['opened', 'closed', 'transition']);

// Deprecate popperOptions in favor of multiple props
if (Object.keys(properties.popperOptions).length > 0) {
	console.warn('@kalimahapps/vue-popper: popperOptions is deprecated. Use the individual props instead.');
}

const { popperInstance, triggerElement, tooltipElement, isOpened, toggleTooltip } = useVuePopper({
	hover: properties.hover,
	disableClickOutside: properties.disableClickOutside,
	openDelay: properties.openDelay,
	closeDelay: properties.closeDelay,
	showArrow: properties.showArrow,
	strategy: properties.strategy,
	modifiers: properties.modifiers,
	animation: properties.animation,
	placement: properties.placement,
	interactive: properties.interactive,
	animationDuration: properties.animationDuration,

	onTransition: (transition: string, element: HTMLElement) => {
		emit('transition', transition, element);
	},
	onOpened: () => {
		emit('opened');
	},
	onClosed: () => {
		emit('closed');
	},
});

defineExpose({
	popper: popperInstance,
	isOpened: computed(() => {
		return isOpened.value;
	}),
	toggleTooltip,
});
</script>
