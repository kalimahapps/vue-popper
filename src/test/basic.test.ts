
import type { VueWrapper } from '@vue/test-utils';
import { mount, flushPromises } from '@vue/test-utils';
import { describe, it, expect, vi } from 'vitest';
import VuePopper from '../VuePopper.vue';
const { fn } = vi;

describe('VuePopper', () => {
	it('should render correctly', () => {
		const wrapper = mount(VuePopper);
		expect(wrapper.html()).toMatchSnapshot();
	});

	it('should clear instance on unmount', () => {
		const wrapper = mount(VuePopper);
		wrapper.unmount();
		expect(wrapper.vm.popperInstance).toBeUndefined();
	});

	it('should change open status on click', async () => {
		const wrapper: VueWrapper<any> = mount(VuePopper);
		await wrapper.find('.trigger-element').trigger('click');
		await flushPromises();
		expect(wrapper.vm.isOpened).toBe(true);
	});

	it('should change open status on hover', async () => {
		const wrapper: VueWrapper<any> = mount(VuePopper, {
			props: {
				hover: true,
				openDelay: 0,
			},
		});

		await wrapper.find('.trigger-element').trigger('mouseenter');
		await flushPromises();

		expect(wrapper.vm.isOpened).toBe(true);
	});

	it('should change not react on mouseenter if hover not enabled', async () => {
		const wrapper: VueWrapper<any> = mount(VuePopper, {
			props: {
				hover: false,
			},
		});

		await wrapper.find('.trigger-element').trigger('mouseenter');
		await flushPromises();
		expect(wrapper.vm.isOpened).toBe(false);
	});

	it('should open and close with 2 seconds delay', async () => {
		const t = fn();

		vi.useFakeTimers();

		const wrapper: VueWrapper<any> = mount(VuePopper, {
			props: {
				openDelay: 2000,
				closeDelay: 2000,
			},
		});

		// Test open delay
		await wrapper.find('.trigger-element').trigger('click');
		setTimeout(t, 2000);
		vi.advanceTimersByTime(2000);
		await flushPromises();
		expect(wrapper.vm.isOpened).toBe(true);

		// Test close delay
		window.document.body.dispatchEvent(new Event('click'));

		setTimeout(fn(), 5000);
		vi.advanceTimersByTime(5000);
		await flushPromises();
		expect(wrapper.vm.isOpened).toBe(false);
		vi.useRealTimers();
	});
});
