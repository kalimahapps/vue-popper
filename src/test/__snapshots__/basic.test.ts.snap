// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VuePopper > should render correctly 1`] = `
"<div class="popper-container">
  <div class="trigger-element" ariadescribedby="tooltip"></div>
  <transition-stub name="fade" appear="false" persisted="false" css="true">
    <div class="popper-content-wrapper" style="display: none; position: absolute; left: 0px; top: 0px; margin: 0px;">
      <transition-stub name="shift-in" appear="false" persisted="false" css="true" data-placement="top" data-name="shift-in">
        <div class="popper-content" style="display: none;">
          <div data-popper-arrow="true" style="display: block;"></div>
        </div>
      </transition-stub>
    </div>
  </transition-stub>
</div>"
`;
